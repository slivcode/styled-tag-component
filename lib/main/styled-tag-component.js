"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var cond_arr_1 = require("../util/cond-arr");
var partition_obj_1 = require("../util/partition-obj");
var map_values_1 = require("../util/map-values");
exports.StyledTagComponent = function (a) {
    var _a = a.tag, tag = _a === void 0 ? 'div' : _a, _b = a.ep, ep = _b === void 0 ? [] : _b, cn = a.cn, _s = a.style, nullIf = a.nullIf;
    return function (p) {
        var _a = partition_obj_1.partitionObj(ep.concat('innerRef'))(p), eP = _a[0], dP = _a[1];
        if (nullIf && nullIf(eP))
            return null;
        var className = cn ? cond_arr_1.condArr(cn(eP)).join(' ') : null;
        var style = _s && map_values_1.mapValues(cond_arr_1.applyIfIsArr(function (v) {
            var t = v.find(function (_a) {
                var b = _a[0];
                return b;
            });
            return (t && t[1]) || null;
        }))(_s(eP));
        return react_1.createElement(tag, __assign({ className: className, style: style, ref: eP.innerRef }, dP));
    };
};
