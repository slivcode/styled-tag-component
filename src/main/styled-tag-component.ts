import { ComponentClass, createElement, CSSProperties, HTMLAttributes, SFC } from 'react';
import { applyIfIsArr, condArr } from '../util/cond-arr';
import { partitionObj } from '../util/partition-obj';
import { mapValues } from '../util/map-values';

export type T_StyledTagArg<P> = {
	tag?: string | any
	ep?: string[]
	cn?: (p: P) => (string | [boolean, string])[],
	style?: (p?: P) => CSSProperties,
	nullIf?: (p?: P) => boolean,
};
export type T_StyledTagComponent = <P>(opt: T_StyledTagArg<P>) => SFC<P & { innerRef?: (el: HTMLElement) => void } & HTMLAttributes<any>>;
export const StyledTagComponent: T_StyledTagComponent = (a) => {
	let { tag = 'div', ep = [], cn, style: _s, nullIf } = a;
	return (p) => {
		let [eP, dP] = partitionObj(ep.concat('innerRef'))(p);
		if (nullIf && nullIf(eP)) return null;
		let className = cn ? condArr(cn(eP)).join(' ') : null;
		let style = _s && mapValues(applyIfIsArr(v => {
				let t = v.find(([b]) => b);
				return (t && t[1]) || null;
			}))(_s(eP));
		return createElement(tag, { className, style, ref: eP.innerRef, ...dP });
	};
};