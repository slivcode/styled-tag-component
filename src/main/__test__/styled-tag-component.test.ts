import { StyledTagComponent } from '../styled-tag-component';
import { mount, shallow } from 'enzyme';
import { createElement } from 'react';

describe('[StyledTagComponent] suite', () => {
	let R = (f) => f();
	test('innerRef is called', () => {
		let E = StyledTagComponent({});
		let f = jest.fn();
		mount(createElement(E, { innerRef: f }));
		expect(f).toBeCalled();
	});
	test('no class attribute if no cn', () => {
		let E = StyledTagComponent({});
		let $ = shallow(createElement(E));
		expect($.html()).not.toMatch('class');
	});
	test('style', () => {
		let E = StyledTagComponent<{ isActive }>({
			ep: ['isActive'],
			style: ({ isActive }) => ({
				color: [
					[isActive, 'red'],
					[true, 'blue'],
				],
				background: 'red',
			}),
		});
		let $ = shallow(createElement(E, { isActive: true }, 'style'));
		expect($.html()).toMatchSnapshot();
		$.setProps({ isActive: false });
		expect($.html()).toMatchSnapshot();
	});
	test('className', () => {
		let E = StyledTagComponent<{ isActive }>({
			ep: ['isActive'],
			cn: ({ isActive }) => [
				[true, 'base'],
				[isActive, 'active'],
			],
		});
		let $ = shallow(createElement(E, { isActive: true }));
		expect($.html()).toMatchSnapshot();
		$.setProps({ isActive: false });
		expect($.html()).toMatchSnapshot();
	});
	test('tagName', () => {
		let E = StyledTagComponent({
			tag: 'div',
		});
		expect(shallow(createElement(E)).html()).toMatchSnapshot();
		let L = StyledTagComponent({
			tag: 'li',
		});
		expect(shallow(createElement(L)).html()).toMatchSnapshot();
	});
	test('ifNull', () => {
		let E = StyledTagComponent<{ isHidden }>({
			ep: ['isHidden'],
			nullIf: ({ isHidden }) => isHidden,
		});
		let $ = shallow(createElement(E));
		expect($.html()).toMatchSnapshot();
		$.setProps({ isHidden: true });
		expect($.html()).toMatchSnapshot();
	});
});