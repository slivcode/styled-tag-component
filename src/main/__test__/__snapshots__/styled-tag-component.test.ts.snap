// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[StyledTagComponent] suite className 1`] = `"<div class=\\"base active\\"></div>"`;

exports[`[StyledTagComponent] suite className 2`] = `"<div class=\\"base\\"></div>"`;

exports[`[StyledTagComponent] suite ifNull 1`] = `"<div></div>"`;

exports[`[StyledTagComponent] suite ifNull 2`] = `null`;

exports[`[StyledTagComponent] suite style 1`] = `"<div style=\\"color:red;background:red;\\">style</div>"`;

exports[`[StyledTagComponent] suite style 2`] = `"<div style=\\"color:blue;background:red;\\">style</div>"`;

exports[`[StyledTagComponent] suite tagName 1`] = `"<div></div>"`;

exports[`[StyledTagComponent] suite tagName 2`] = `"<li></li>"`;
